java/ql/src/experimental/Security/CWE/CWE-078/CommandInjectionRuntimeExec.qll

```diff
@@ -0,0 +1,170 @@
+}
+
+// taint flow from user data to args of the command
+deprecated class ExecTaintConfiguration2 extends TaintTracking::Configuration {
+  ExecTaintConfiguration2() { this = "ExecTaintConfiguration2" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(RuntimeExecMethod method, MethodAccess call |
+      call.getMethod() = method and
+      sink.asExpr() = call.getArgument(_) and
+      sink.asExpr().getType() instanceof Array
+    )
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.asExpr().getFile().isSourceFile() and
+    (
+      node.getType() instanceof PrimitiveType or
+      node.getType() instanceof BoxedType
+    )
+  }
+}
+
+module ExecUserFlowConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    source instanceof Source
+  }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(RuntimeExecMethod method, MethodAccess call |
+      call.getMethod() = method and
+      sink.asExpr() = call.getArgument(_) and
+      sink.asExpr().getType() instanceof Array
+    )
+  }
+
+  predicate isBarrier(DataFlow::Node node) {
+    node.asExpr().getFile().isSourceFile() and
+    (
+      node.getType() instanceof PrimitiveType or
+      node.getType() instanceof BoxedType
+    )
+  }
+}
+
+/** Tracks flow of unvalidated user input that is used in Runtime.Exec */
+module ExecUserFlow = TaintTracking::Global<ExecUserFlowConfig>;
+
+// array[3] = node
+class AssignToNonZeroIndex extends DataFlow::Node {
+  AssignToNonZeroIndex() {
+    exists(AssignExpr assign, ArrayAccess access |
+      assign.getDest() = access and
+      access.getIndexExpr().(IntegerLiteral).getValue() != "0" and
+      assign.getSource() = this.asExpr()
+    )
+  }
+}
+
+// String[] array = {"a", "b, "c"};
+class ArrayInitAtNonZeroIndex extends DataFlow::Node {
+  ArrayInitAtNonZeroIndex() {
+    exists(ArrayInit init, int index |
+      init.getInit(index) = this.asExpr() and
+      index != 0
+    )
+  }
+}
+
+// Stream.concat(Arrays.stream(array_1), Arrays.stream(array_2))
+class StreamConcatAtNonZeroIndex extends DataFlow::Node {
+  StreamConcatAtNonZeroIndex() {
+    exists(MethodAccess call, int index |
+      call.getMethod().getQualifiedName() = "java.util.stream.Stream.concat" and
+      call.getArgument(index) = this.asExpr() and
+      index != 0
+    )
+  }
+}
+
+// allow list of executables that execute their arguments
+// TODO: extend with data extensions
+class UnSafeExecutable extends string {
+  bindingset[this]
+  UnSafeExecutable() {
+    this.regexpMatch("^(|.*/)([a-z]*sh|javac?|python.*|perl|[Pp]ower[Ss]hell|php|node|deno|bun|ruby|osascript|cmd|Rscript|groovy)(\\.exe)?$") and
+    not this = "netsh.exe"
+  }
+}
+
+predicate callIsTaintedByUserInputAndDangerousCommand(MethodAccess call, ExecUserFlow::PathNode source, ExecUserFlow::PathNode sink, DataFlow::Node sourceCmd, DataFlow::Node sinkCmd) {
```

Comment:
You're actually not using `call` anywhere outside of this predicate, so you could remove it from the parameters and add it as an `exists` variable.